### PR TITLE
Address unsynchronized reset issue

### PR DESCRIPTION
**Pull request by codemercenary**
This pull request only partially addresses the asynchronous reset problem.  There is no general solution to this problem because at the heart of the issue is a cancellation race with `NotifyWhenAutowired`.  The simplest case that demonstrates this issue is as follows:

Thread 1:
```C++
{
    std::lock_guard<std::mutex> lk(lock);
    Autowired<T> ptr;
    ptr->NotifyWhenAutowired([&] {
        std::lock_guard<std::mutex> lk(lock);
        // ...do more work...
    });
    std::this_thread::sleep_for(std::chrono::seconds(1));
}
```

If `T` is injected during the sleep, the callback will be invoked and the injecting thread will block on the lock acquisition attempt.  The sleeping thread wakes up, and tries to destroy `ptr`.  This causes `DeferrableAutowiring::~DeferrableAutowiring()` to be called.

1.  If this destructor blocks until all registered handlers have been cancelled, a deadlock results.
2.  If it cancels on a best-effort basis, a potential use-after-free results.

The only solution is to require that users externally synchronize `Autowired::~Autowired` with `CoreContext::Inject`.

### CODE REVIEW

File: autowiring/CoreContext.h [hunk 1]
```diff
@@ -1137,75 +1137,6 @@ class CoreContext:
...
-
-  /// <summary>
-  /// Registers a slot to be autowired
-  /// </summary>
-  /// <param name="ref">The space where the resolved shared pointer will be stored</param>
-  /// <param name="defer">
-  /// In the event that resolution was not successful, the deferrable that will be used to perform delayed satisfaction
-  /// </param>
-  template<class T>
-  void Autowire(AnySharedPointerT<T>& ref, DeferrableAutowiring& defer) {
```
Comment (jdonald) [inline autowiring/CoreContext.h]: This function is still called on line 105 of autowiring/Autowired.h, leading to a build error on all systems. Removing that clause reveals other compilation issues.

Comment (codemercenary) [inline autowiring/CoreContext.h]: This PR is still underway.

Comment (jdonald) [inline autowiring/CoreContext.h]: Sorry about that. I got too excited and jumped the gun on attempting to test this w/ unit tests/DumpReplay.

File: autowiring/CoreContext.h [hunk 2]
```diff
@@ -670,6 +622,16 @@ class CoreContext:
...
   }
 
+  /// \internal
+  /// <summary>
+  /// Scans the memo collection for the specified entry, or adds a deferred resolution marker if resolution was not possible
+  /// </summary>
+  /// <returns>
+  /// The memo entry where this type was found
+  /// </returns>
+  /// <param name="id">The type to be located</param>
```
Comment (veronicaz41) [inline autowiring/CoreContext.h]: Should be `<param name="type">`

File: src/autowiring/CoreContext.cpp [hunk 3]
```diff
@@ -797,8 +795,7 @@ void CoreContext::UnregisterEventReceiversUnsafe(void) {
     m_pParent->RemoveEventReceivers(m_eventReceivers);
 
   // Recursively unregister packet factory subscribers:
-  AnySharedPointerT<AutoPacketFactory> pf;
-  FindByTypeUnsafe(pf);
+  FindByTypeUnsafe(auto_id_t<AutoPacketFactory>{});
```
Comment (veronicaz41) [inline src/autowiring/CoreContext.cpp]: How does this unregister packet factory subscribers?

Comment (codemercenary) [inline src/autowiring/CoreContext.cpp]: Good freaking question...let me review this.

Comment (codemercenary) [inline src/autowiring/CoreContext.cpp]: It doesn't.  This is handled in `AutoPacketFactory::OnStop`

File: src/autowiring/AutowirableSlot.cpp [hunk 4]
```diff
@@ -8,45 +8,73 @@
...
+  // that we perform an expensive memory reallocation, so let's pay a little memory in advance to help make the
+  // most common use case pretty fast.
+  m_autowired_notifications.reserve(2);
+  m_autowired_notifications.emplace_back(std::move(reg));
 }
 
-void DeferrableAutowiring::RegisterDeferredAutowire(autowiring::registration_t&& reg) {
-  m_deferred_registration = std::move(reg);
+DeferrableAutowiring::~DeferrableAutowiring(void) {
+  // BUG:  The reset call does not strictly prohibit
```
Comment (veronicaz41) [inline src/autowiring/AutowirableSlot.cpp]: Why this is a bug?

Comment (codemercenary) [inline src/autowiring/AutowirableSlot.cpp]: Forgot to explain this fully.

File: src/autowiring/AutowirableSlot.cpp [hunk 5]
```diff
@@ -8,45 +8,80 @@
...
+  // Local versions of the members we are clearing out
+  std::weak_ptr<CoreContext> contextWeak;
+  AnySharedPointer ptr;
+  std::vector<autowiring::registration_t> autowired_notifications;
+  
+  // Move and handle resources under lock
+  {
+    std::lock_guard<autowiring::spin_lock> lk{ m_lock };
+    autowired_notifications = std::move(m_autowired_notifications);
+    ptr = std::move(m_ptr);
```
Comment (veronicaz41) [inline src/autowiring/AutowirableSlot.cpp]: Why we need to move `m_ptr` here?

Comment (veronicaz41) [inline src/autowiring/AutowirableSlot.cpp]: OK. this is to make sure `m_ptr` is released.

File: src/autowiring/CoreContext.cpp [hunk 6]
```diff
@@ -300,81 +301,78 @@ void CoreContext::AddInternal(const AnySharedPointer& ptr) {
...
     throw autowiring_error("This interface is already present in the context");
 
   // Now we can satisfy it:
+  entry.m_local = true;
   entry.m_value = ptr;
   UpdateDeferredElement(std::move(lk), entry);
 }
 
-MemoEntry& CoreContext::FindByType(AnySharedPointer& reference, bool localOnly) const {
+MemoEntry& CoreContext::FindByType(auto_id type, bool nonrecursive) const {
```
Comment (jdonald) [inline src/autowiring/CoreContext.cpp]: This breaks compatibility with an existing use case.

    std::shared_ptr<WinUSBPayloadReader> rdr;
    streaming->FindByType<WinUSBPayloadReader>(rdr);

Given we can't use auto_id there, what is the replacement? Also, would this push the version to 0.8.0?

Comment (codemercenary) [inline src/autowiring/CoreContext.cpp]: I'll add a variant that will restore compatibility.

Comment (codemercenary) [inline src/autowiring/CoreContext.cpp]: Are you sure this breaks compatibility?  `CoreContext.h:1107` provides the shared pointer version of this routine.

**Pull request merged**